File: 135.分发糖果.java
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=135 lang=java
 *
 * [135] 分发糖果
 */

// @lc code=start
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
        if (n < 2)  return n;
        int[] candies = new int[n];
        Arrays.fill(candies, 1);
        
        // iterate from left to right
        for (int i = 1; i < n; ++i) {
            if (ratings[i] > ratings[i-1]) {
                candies[i] = candies[i-1] + 1;
            }
        }

        // then iterate from right to left and sum things up at the same time
        int sum = candies[n-1];
        for (int i = n - 2; i >= 0; --i) {
            if (ratings[i] > ratings[i+1] && candies[i] < candies[i+1] + 1) {
                candies[i] = candies[i+1] + 1;
            }
            sum += candies[i];
        }
        return sum;
    }
}
// @lc code=end

File: 135.分发糖果.py
#
# @lc app=leetcode.cn id=135 lang=python3
#
# [135] 分发糖果
#

# @lc code=start
class Solution:
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        if n < 2:
            return n
        candies = [1] * n

        # 先从左往右遍历一遍，如果右边孩子的评分比左边的高，则右边孩子的糖果数更新为
        # 左边孩子的糖果数加1。
        for i in range(1, n):
            if ratings[i] > ratings[i - 1]:
                candies[i] = candies[i - 1] + 1

        # 再从右往左遍历一遍，如果左边孩子的评分比右边的高，且左边孩子当前的糖果数
        # 不大于右边孩子的糖果数，则左边孩子的糖果数更新为右边孩子的糖果数加1。
        # 与此同时，第二次遍历可以顺带求和。
        ans = candies[-1]
        for i in range(n - 2, -1, -1):
            if ratings[i] > ratings[i + 1]:  # and candies[i] < candies[i + 1] + 1:
                candies[i] = max(candies[i], candies[i + 1] + 1)
            ans += candies[i]
        return ans

# @lc code=end

